[#52] - Posts DTO

Dj5822 opened the pull request:
# Description:
Edited the model such that two new table are created: likedPost and sharedPost.
Removed the attachment field in the post table. This is because we may not necessarily need this field depending on how the attachment is implemented (we might create another table instead).
Created the post DTO based on the specifications required.
Note that the usersLiked, usersShared, and attachment is still a work in progress.

### All Submissions:

* [x] Have you followed the guidelines in our [Contributing document](../CONTRIBUTING.md)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../../pulls) for the same update/change?

### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] My code follows the code style of this project
- [x] My code has been commented
- [ ] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly
- [x] I have added tests to cover my changes
- [x] All new and existing tests passed

Dj5822 committed:
Updated the database

Created two new tables: likedpost and sharedpost to store the associations between posts and users.

Dj5822 committed:
Merge branch 'main' of https://github.com/Dj5822/Updog into posts-dto

Dj5822 committed:
Added dto for posts

Uses dto for POST and GET method.

Dj5822 committed:
Merge branch 'main' of https://github.com/Dj5822/Updog into posts-dto

hmit106 reviewed (commented):
Have you considered what unit tests you could add to verify DTO functionality?

mdar365 commented on backend/server/controllers/posts.js:
```diff
@@ -166,7 +164,7 @@ export const deletePostById = async (req, res) => {
                         res.status(200).send('The message has been deleted.')
                     } else {
-                        res.status(400).send('Failed to update.')
+                        res.status(400).send('Failed to delete.')
```
I think this should be 500, because if we failed to delete it for an unknown reason then it's a server error rather than a client error

Also I just realised, the if statement above it should say "The post has been deleted" right?

Dj5822 replied:
Actually, I think 400 is correct. It's the error message that is incorrect. The error occurs when the id for that particular post can not be found.

Dj5822 replied:
Wait nvm, that should be 404 not found.

mdar365 replied:
But you already checked if the post exists when you typed `if (!post) {`

Dj5822 replied:
yeah i just realised

Dj5822 replied:
I made the change now

mdar365 commented on backend/dto/posts.js:
```diff
@@ -0,0 +1,22 @@
+        children: children.map((child) => child.id),
+        usersLiked: 0, // work in progress.
+        usersShared: 0, // work in progress.
+        timestamp: post.updatedAt,
```
front-end expects the timestamp to be returned as a unix timestamp, updatedAt is of type date so it should be converted first

Dj5822 replied:
I think I fixed it.

mdar365 commented on backend/dto/posts.js:
```diff
@@ -0,0 +1,22 @@
+        author,
+        parent: post.parent,
+        children: children.map((child) => child.id),
+        usersLiked: 0, // work in progress.
+        usersShared: 0, // work in progress.
```
I recommend finishing that part first before merging the PR

Dj5822 replied:
Fixed it. Now it shows the actual number of likes and shares. Had to fix the model as well.

mdar365 reviewed (changes requested):
Just some minor changes, I really like how you created separate tables for liked and shared posts because this will really help us when we're creating the feed and user profile

We should also modify the post endpoint tests to expect a postDTO

Dj5822 committed:
Updated error codes

Updated the error codes and message such that 500 is sent instead of 400 if the database modification fail.

Dj5822 committed:
Updated usersLiked and usersShared

Changed usersLiked and usersShared in the posts dto such that they return the actual number of likes and shares.

Dj5822 committed:
Merge branch 'main' of https://github.com/Dj5822/Updog into posts-dto

Dj5822 committed:
Converted datatypes to integers

Dj5822 committed:
Fixed typo

Dj5822 committed:
Modified tests to check the contents of the dto

Checks certain values such as the content, author's username and email, parent, children, usersLiked, and usersShared.

Dj5822 committed:
Improved tests further

Now checks whether the id exists and whether the timestamp exists. Doesn't check whether these values are valid though.

Dj5822 commented:
Ok I think I'm done now.
